File: cpp/dp/main_1699.cpp
#include <iostream>
#include <cmath>

using namespace std;

int main() {
	int num;
	cin >> num;

	int* dp = new int[num + 1];
	dp[0] = 0;
	dp[1] = 1;
	for (int i = 2; i <= num; i++) {
		int s = static_cast<int>(sqrt(i));
		if (s * s == i) {
			dp[i] = 1;
		}
		else {
			dp[i] = dp[i - 1] + 1;
			for (int j = 2; j < s; ++j) {
				dp[i] = min(dp[i], dp[i - j * j] + 1);
			}
		}
	}
	cout << dp[num] << endl;
	delete[] dp;
}

File: kotlin/dp/main_1699.kt
import kotlin.math.min
import kotlin.math.sqrt

fun main() {
    val num = readLine()!!.toInt()
    val dp = MutableList(num + 1) { it }
    for (i in 3..num){
        val s = sqrt(i.toDouble()).toInt()
        if (s * s == i)
            dp[i] = 1
        else
        for (j in 1..s)
            dp[i] = min(dp[i], dp[i - j * j] + 1)
    }
    println(dp[num]);
}